Add authenticated endpoint to download laravel log

pjaudiomv opened the pull request:
Should this be restricted to server admin?

alanb2718 commented:
Regarding whether this should be restricted to the server admin: will the log file ever contain information that shouldn't be viewed by a random service body admin, for example information about other service bodies or contacts?  If so, then yes, let's restrict it.  (I'm not sure why anyone other than the server admin would want to download it anyway, but if there isn't confidential information then it doesn't seem like a necessary restriction.)

jbraswell reviewed (changes requested):
I think logs/laravel would be better.

Needs swagger annotations so the client can have it.

Needs a failing test with an unauthenticated user and with other user types.

Probably needs policies to enforce permissions?

pjaudiomv commented:
I changed to logs/laravel , added swagger and restricted to server admin. as well as tests for unauthenticated and service body admin

<img width="1341" alt="Screenshot 2024-08-17 at 7 59 54 PM" src="https://github.com/user-attachments/assets/a4f0873d-a534-4a32-acff-7d1a4041d8e3">

jbraswell commented on src/app/Http/Controllers/Admin/LaravelLogController.php:
```diff
@@ -0,0 +1,28 @@
+{
+    public function laravelLog(Request $request): JsonResponse|BinaryFileResponse
+    {
+        $user = $request->user();
```
Permissions should be enforced using a policy. Look at app/Policies.

jbraswell replied:
Actually, I think policies only apply if you're dealing with models? It's worth looking into, but your code may be fine as written.

pjaudiomv replied:
Ya I looked into it, it only applies to models

jbraswell commented on src/app/Http/Controllers/Admin/Swagger/LaravelLogController.php:
```diff
@@ -0,0 +1,29 @@
+     *         @OA\JsonContent(ref="#/components/schemas/AuthenticationError")
+     *     ),
+     *     @OA\Response(response=403, description="Returns when user is unauthorized to perform action.",
+     *         @OA\JsonContent(ref="#/components/schemas/AuthenticationError")
```
This should be AuthorizationError.

pjaudiomv replied:
fixed

jbraswell commented on src/storage/api-docs/api-docs.json:
```diff
@@ -486,6 +486,63 @@
+                        "content": {
+                            "application/json": {
+                                "schema": {
+                                    "$ref": "#/components/schemas/AuthenticationError"
```
Should be authorization instead of authentication.

pjaudiomv replied:
fixed

jbraswell commented on src/app/Http/Controllers/Admin/LaravelLogController.php:
```diff
@@ -0,0 +1,28 @@
+use Illuminate\Http\JsonResponse;
+use Illuminate\Support\Facades\Gate;
+
+class LaravelLogController extends Controller
```
This controller should probably be LogController, to mirror /logs. Then, the method should probably be called laravel, to mirror the "laravel" part of /logs/laravel.

In rest language, logs is the resource type, and laravel is specific instance of that resource.

pjaudiomv replied:
thanks just altered the things accordingly

jbraswell commented on src/app/Http/Controllers/Admin/LaravelLogController.php:
```diff
@@ -0,0 +1,28 @@
+        $filePath = storage_path('logs/laravel.log');
+
+        if (file_exists($filePath)) {
+            return response()->download($filePath);
```
Does laravel have a good way to gzip the outgoing bytes?

pjaudiomv replied:
didnt see anything magical with like arrows and shit

jbraswell commented on src/app/Http/Controllers/Admin/Swagger/LaravelLogController.php:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * @OA\Get(path="/api/v1/logs/laravel", summary="Retrieves laravel log", description="Retrieve the laravel log if it exists.", operationId="getLaravelLog", tags={"rootServer"}, security={{"bmltToken":{}}},
+     *     @OA\Response(response=200, description="Returns when user is authenticated.",
+     *         @OA\MediaType(mediaType="text/plain", @OA\Schema(type="string"))
```
Look at the "Response That Returns a File" section of https://swagger.io/docs/specification/describing-responses/.

pjaudiomv replied:
fixed, its cool the swagger ui generates a download link for you with the gzip file

jbraswell commented on src/app/Http/Controllers/Admin/LogController.php:
```diff
@@ -0,0 +1,33 @@
+        $filePath = storage_path('logs/laravel.log');
+
+        if (file_exists($filePath)) {
+            $fileContent = file_get_contents($filePath);
```
Rather than reading the whole file into memory, try reading it in chunks. Then you can use streamDownload instead of just download to stream the file back to the client and never have to worry about out of memory errors if the log file is abnormally large. Take a look at the getNawsDump code for an example.

pjaudiomv replied:
done

jbraswell commented on src/app/Http/Controllers/Admin/Swagger/LogController.php:
```diff
@@ -0,0 +1,29 @@
+     *     ),
+     * )
+     */
+    public function index()
```
For consistency this should have the same name as the method in LogController, unless it breaks something.

pjaudiomv replied:
done

jbraswell commented on src/tests/Feature/Admin/LogTest.php:
```diff
@@ -0,0 +1,77 @@
+
+    public function testLaravelLogDownloadSuccess()
+    {
+        if (!File::exists(storage_path('logs'))) {
```
You should avoid mutating anything real in a test. In this case, that's the filesystem. See https://phpunit.de/manual/6.5/en/test-doubles.html#test-doubles.mocking-the-filesystem for information on how to mock the filesystem.

pjaudiomv replied:
I had tried to use `Storage::fake()` and that didnt work and the other lib they suggested looks like its not maintained anymore so didnt wanna add it as a dep.

jbraswell replied:
I don't feel comfortable approving a test that mutates the filesystem.

pjaudiomv replied:
I commented out the offending two tests for now and added a todo with note

jbraswell commented on src/tests/Feature/Admin/LogTest.php:
```diff
@@ -0,0 +1,77 @@
+            ->assertHeader('Content-Disposition', 'attachment; filename=laravel.log.gz');
+    }
+
+    public function testLogDownloadNoLogFailure()
```
The test name should ideally indicate why it failed. In this case, the name should include NotFound.

pjaudiomv replied:
done
